parlai/core/torch_generator_agent.py

```diff
@@ -1090,163 +891,96 @@ def get_hyp_from_finished(self, hypothesis_tail):
+        """
+        # if we never actually finished, force one
+        if not self.finished:
+            self.finished.append(
+                _HypothesisTail(
+                    timestep=len(self.outputs) - 1,
+                    hypid=0,
+                    score=self.all_scores[-1][0],
+                    tokenid=self.eos,
+                )
+            )
+
         rescored_finished = []
         for finished_item in self.finished:
             current_length = finished_item.timestep + 1
             # these weights are from Google NMT paper
             length_penalty = math.pow((1 + current_length) / 6, 0.65)
             rescored_finished.append(
-                self.HypothesisTail(
+                _HypothesisTail(
                     timestep=finished_item.timestep,
                     hypid=finished_item.hypid,
                     score=finished_item.score / length_penalty,
                     tokenid=finished_item.tokenid,
                 )
             )
 
+        # Note: beam size is almost always pretty small, so sorting is cheap enough
         srted = sorted(rescored_finished, key=attrgetter('score'), reverse=True)
 
         if n_best is not None:
             srted = srted[:n_best]
 
-        return srted
+        return [
+            (self._get_pretty_hypothesis(self._get_hyp_from_finished(hyp)), hyp.score)
+            for hyp in srted
+        ]
 
-    def check_finished(self):
-        """
-        Check if self.finished is empty and add hyptail in that case.
 
-        This will be suboptimal hypothesis since the model did not get any EOS
-        """
-        if len(self.finished) == 0:
-            # we change output because we want outputs to have eos
-            # to pass assert in L102, it is ok since empty self.finished
-            # means junk prediction anyway
-            self.outputs[-1][0] = self.eos
-            hyptail = self.HypothesisTail(
-                timestep=len(self.outputs) - 1,
-                hypid=0,
-                score=self.all_scores[-1][0],
-                tokenid=self.outputs[-1][0],
-            )
+class GreedySearch(TreeSearch):
+    """
+    Greedy search.
 
-            self.finished.append(hyptail)
+    Picks the highest probability utterance at each step.  Only works with
+    --beam-size 1.
+    """
 
-    def get_beam_dot(self, dictionary=None, n_best=None):
-        """
-        Create pydot graph representation of the beam.
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.beam_size != 1:
+            raise ValueError('Greedy search can only be run with beam size 1.')
 
-        :param outputs:
-            self.outputs from the beam
+    def select_paths(self, logprobs, prior_scores):
+        s, i = logprobs.max(1)
+        best_scores = s + prior_scores.unsqueeze(1)
+        hyp_ids = torch.arange(logprobs.size(0)).to(logprobs.device)
+        return (hyp_ids, i, best_scores)
```

Comment:
is `i` a list here? in the `select_paths` docstring above it mentions that this will be a list, just curious if this is/ if this will break things if not

Reply:
better variable names, yes (y)